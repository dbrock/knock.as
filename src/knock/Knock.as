package knock {
  public class Knock {
    public static var equal: Function = knock.equal
  }
}
