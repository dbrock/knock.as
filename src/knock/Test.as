package knock {
  import stdio.flash.Sprite
  import stdio.process

  public class Test extends Sprite {
    override public function main(): void {
      run()
      process.exit()
    }

    public function run(): void {}
  }
}
