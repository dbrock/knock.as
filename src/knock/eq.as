package knock {
  public function eq(a: *, b: *, desc: String = null): void {
    ok(a == b, desc)
  }
}
