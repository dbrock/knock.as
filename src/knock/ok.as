package knock {
  import stdio.process

  public function ok(passed: Boolean, desc: String = null): void {
    desc = desc === null ? "" : " - " + desc

    if (passed) {
      process.puts("ok" + desc)
    } else {
      process.puts("not ok" + desc)
    }
  }
}
