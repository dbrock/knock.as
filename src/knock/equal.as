package knock {
  public function equal(x: Object, y: Object): Boolean {
    if (x && ("equals" in x)) {
      return x.equals(y)
    } else if (y && ("equals" in y)) {
      return y.equals(x)
    } else if (x is Array && y is Array) {
      return arrays_equal(x as Array, y as Array)
    } else if (is_plain_object(x) && is_plain_object(y)) {
      return objects_equal(x, y)
    } else {
      return x == y
    }
  }
}

import knock.*

function arrays_equal(x: Array, y: Array): Boolean {
  if (x.length === y.length) {
    for (var i: int = 0; i < x.length; ++i) {
      if (!equal(x[i], y[i])) {
        return false
      }
    }

    return true
  } else {
    return false
  }
}

function is_plain_object(value: Object): Boolean {
  return value !== null && value.toString() === "[object Object]"
}

function objects_equal(x: Object, y: Object): Boolean {
  const x_keys: Array = get_object_keys(x)
  const y_keys: Array = get_object_keys(y)

  x_keys.sort()
  y_keys.sort()

  if (equal(x_keys, y_keys)) {
    for each (var key: String in x_keys) {
      if (!equal(x[key], y[key])) {
        return false
      }
    }

    return true
  } else {
    return false
  }
}

function get_object_keys(object: Object): Array {
  const result: Array = []

  for (var key: String in object) {
    result.push(key)
  }

  return result
}
