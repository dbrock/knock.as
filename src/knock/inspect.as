package knock {
  import flash.utils.getQualifiedClassName

  public function inspect(object: Object): String {
    if (object === null) {
      return "null"
    } else if (typeof object === "number") {
      return object.toString()
    } else if (typeof object === "string") {
      return inspect_string(object as String)
    } else if (object is Array) {
      return inspect_array(object as Array)
    } else if (object.toString() === "[object Object]") {
      return inspect_object(object)
    } else if (/^\[object \w+\]$/.test(object.toString())) {
      return "#<" + getQualifiedClassName(object) + ">"
    } else {
      return object.toString()
    }
  }
}

import knock.*

function inspect_string(string: String): String {
  return '"' + escape_string(string) + '"'
}

function escape_string(string: String): String {
  return string
    .replace(/ \\ /gx, "\\\\")
    .replace(/"/g, '\\"')
}

function inspect_array(array: Array): String {
  const parts: Array = []

  for each (var element: Object in array) {
    parts.push(inspect(element))
  }

  return "[" + parts.join(", ") + "]"
}

function inspect_object(object: Object): String {
  const items: Array = []

  for (var name: String in object) {
    items.push(name + ": " + inspect(object[name]))
  }

  if (items.length === 0) {
    return "{}"
  } else {
    return "{ " + items.join(", ") + " }"
  }
}
