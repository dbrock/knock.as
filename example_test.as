package {
  import knock.*

  public class example_test extends Test {
    override public function run(): void {
      ok(true)
      ok(true, "a")
      eq(1, 1, "b")
      eq(1, 2)
    }
  }
}
