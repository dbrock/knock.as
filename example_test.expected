ok
ok - a
ok - b
not ok
